geps/gep-1713/index.md

```diff
@@ -0,0 +1,381 @@
+type ChildrenNamespaces struct {
+	// From indicates where Children will be selected for this Gateway. Possible
+	// values are:
+	//
+	// * All: Children in all namespaces may be used by this Gateway.
+	// * Selector: Children in namespaces selected by the selector may be used by
+	//   this Gateway.
+	// * Same: Only Children in the same namespace may be used by this Gateway.
+	//
+	// Support: Core
+	//
+	// +optional
+	// +kubebuilder:default=Same
+	From *FromNamespaces `json:"from,omitempty"`
+
+	// Selector must be specified when From is set to "Selector". In that case,
+	// only Children in Namespaces matching this Selector will be selected by this
+	// Gateway. This field is ignored for other values of "From".
+	//
+	// Support: Core
+	//
+	// +optional
+	Selector *metav1.LabelSelector `json:"selector,omitempty"`
+}
+
+// GatewayObjectReference identifies an API object including its namespace,
+// defaulting to Gateway.
+type GatewayObjectReference struct {
+	// Group is the group of the referent. For example, "gateway.networking.k8s.io".
+	// When unspecified or empty string, core API group is inferred.
+	//
+	// +optional
+	// +kubebuilder:default=""
+	Group *Group `json:"group"`
+
+	// Kind is kind of the referent. For example "Gateway".
+	//
+	// +optional
+	// +kubebuilder:default=Gateway
+	Kind *Kind `json:"kind"`
+
+	// Name is the name of the referent.
+	Name ObjectName `json:"name"`
+
+	// Namespace is the namespace of the referenced object. When unspecified, the local
+	// namespace is inferred.
+	//
+	// Support: Core
+	//
+	// +optional
+	Namespace *Namespace `json:"namespace,omitempty"`
+}
+```
+
+#### YAML
+
+Below shows an example of an end to end configuration with Gateway merging.
+Here we define a parent resource, which allows children from the same namespace.
+A single listener is defined in the parent.
+
+The child Gateway attaches to this Gateway and specifies an additional listener.
+
+```yaml
+apiVersion: gateway.networking.k8s.io/v1beta1
+kind: Gateway
+metadata:
+  name: parent-gateway
+spec:
+  gatewayClassName: example
+  infrastructure:
+    allowedChildren:
+      namespaces:
+        from: Same
+  listeners:
+  - name: common-monitoring
+    port: 8080
+    protocol: HTTP
+---
+apiVersion: gateway.networking.k8s.io/v1beta1
+kind: Gateway
+metadata:
+  name: child-gateway
+spec:
+  gatewayClassName: example
+  infrastructure:
+    attachTo:
+      name: parent-gateway
+      kind: Gateway
+      group: gateway.networking.k8s.io
+  listeners:
+  - name: domain-a
+    hostname: a.example.com
+    protocol: HTTP
+    port: 80
+```
+
+Logically, this is equivalent to a single Gateway as below:
+
+```yaml
+apiVersion: gateway.networking.k8s.io/v1beta1
+kind: Gateway
+metadata:
+  name: gateway
+spec:
+  gatewayClassName: example
+  listeners:
+  - name: common-monitoring
+    port: 8080
+    protocol: HTTP
+  - name: domain-a
+    hostname: a.example.com
+    protocol: HTTP
+    port: 80
```

Comment:
Agreed, given the following things, which seem on the face of it to be reasonable asks:
- Gateway merging merges the Listeners of each Gateway into one flat list, essentially
- Gateway merging can occur across namespaces

There's no way around the fact that conflicts are not predictable by looking at any single namespace plus the parent Gateway. There's no way to know if someone in a sibling Gateway in another namespace is already using the same port/protocol/hostname/TLS config you are, which will put those Listeners into conflict.

By the current rules, Listeners that are conflicting are _both_ rejected, because they must necessarily be created at the same time (since they are in the same object). Assuming that we use our usual conflict-resolution rules, then, as @robscott says above, the first Gateway to claim the any distinct Listener combination will be the winner forever.

Reply:
>  Unfortunately this doesn't provide any kind of controls for that so it seems like the first namespace to claim a domain would win. A key idea of Gateway listeners was that they allowed you to delegate domains to routes in different namespaces, this seems to make that distinction less clear.

Not really - it's really more about splitting up the listener list across resources.

> A key idea of Gateway listeners was that they allowed you to delegate domains to routes in different namespaces, this seems to make that distinction less clear.

This doesn't change that in my opinion.

> Are there any controls/guardrails we can offer here? Is cross-namespace delegation a key goal? If not, could/should we start without it?

@howardjohn @mikemorris you have comments here regarding the cross namespace need?

Reply:
> There's no way to know if someone in a sibling Gateway in another namespace is already using the same port/protocol/hostname/TLS config you are, which will put those Listeners into conflict.

This isn't really any different than what we have now. eg. if Gateway's aren't visible to Application Developers then they wouldn't know. I agree with your point it becomes more difficult when the resource is split up.

> By the current rules, Listeners that are conflicting are both rejected, because they must necessarily be created at the same time (since they are in the same object). Assuming that we use our usual conflict-resolution rules, then, as @robscott says above, the first Gateway to claim the any distinct Listener combination will be the winner forever.

The GEP tweaks the rules such that child Gateways can be rejected and it doesn't affect siblings. So like you said conflict resolution rules apply.

> As @robscott says above, the first Gateway to claim the any distinct Listener combination will be the winner forever.

Yup.

Reply:
> Is cross-namespace delegation a key goal? If not, could/should we start without it?

I think there's two aspects to this:
- Should a child Gateway be able to attach to a parent Gateway in a different namespace?
  - I think this capability was somewhat implied by the discussion around `allowedChildren` up in https://github.com/kubernetes-sigs/gateway-api/pull/1863#discussion_r1540041898, but we should probably consider this directly - it may add more complexity than value.
- When a Gateway with attached routes attaches to a parent Gateway, what permissions are required on the parent? Does `allowedRoutes` config on the _parent_ Gateway apply here, such as if the parent specified `allowedRoutes` with a selector for only a _different_ namespace? Should we specify `allowedRoutes` as scoped to only _directly_ attached routes?

Reply:
I think for at least some of the use cases, cross-namespace attachment seems important - which implies that we should have a two-way handshake for safety. Which is difficult with the current object. Sigh. Another reason that @robscott and I have been talking about a new object type instead.